### DISCUSSION TITLE
Aeotec Window/Door Sensor Pro 7 / ZWA012 HA Issue

### DISCUSSION THREAD

**Discussion by jdiegmueller** [zwave-js/zwave-js#5453] [General]
Hello!  I'm having trouble with Aeotec Window/Door Sensor Pro 7 / ZWA012 sensors, and am not sure which component to report the problem on (ZwaveJS vs Homeassistant).  Could someone point me in the right direction?  At the moment I'm running ZWaveJS 10.10.0+ZWaveJS-UI 8.8.5 on a Pi4, and HA 2023.2.3.  HA talks to ZWaveJS via WS integration.

So I have fourteen of these ZWA012s, but I am only having trouble with the newer sensors that run the 2.x firmware.  Aeotec's firmware page has a warning advising that there are two different hardware revisions, and that 1.x and 2.x are not compatible with one another.  

The issue I'm having is that the sensors running 2.x do not reflect open/closed in Home Assistant on the typical "window/door is open" entity.

When a 1.x device opens, ZWaveJS logs the following + the "window/door is open" entity changes to open:
```2023-02-11T00:07:43.080Z SERIAL « 0x011200a800018c097105000000ff06160000a4fe                          (20 bytes)
2023-02-11T00:07:43.082Z SERIAL » [ACK]                                                                   (0x06)
2023-02-11T00:07:43.084Z CNTRLR   [Node 140] [~] [Notification] alarmType: 0 => 0                   [Endpoint 0]
2023-02-11T00:07:43.091Z CNTRLR   [Node 140] [~] [Notification] alarmLevel: 0 => 0                  [Endpoint 0]
2023-02-11T00:07:43.095Z DRIVER « [Node 140] [REQ] [BridgeApplicationCommand]
                                  │ RSSI: -92 dBm
                                  └─[NotificationCCReport]
                                      notification type:   Access Control
                                      notification status: 255
                                      notification state:  Window/door is open
2023-02-11T00:07:43.098Z CNTRLR   [Node 140] [~] [Notification] Access Control[Door state]: 23 => 2 [Endpoint 0]
                                  2
```

When a 2.x device opens, ZWaveJS logs the following + the "window/door is open" entity changes is unchanged:
```2023-02-11T00:09:12.657Z SERIAL « 0x011300a80001c00a7105000000ff0616010000a8bd                        (21 bytes)
2023-02-11T00:09:12.659Z SERIAL » [ACK]                                                                   (0x06)
2023-02-11T00:09:12.662Z CNTRLR   [Node 192] [~] [Notification] alarmType: 0 => 0                   [Endpoint 0]
2023-02-11T00:09:12.666Z CNTRLR   [Node 192] [~] [Notification] alarmLevel: 0 => 0                  [Endpoint 0]
2023-02-11T00:09:12.669Z DRIVER « [Node 192] [REQ] [BridgeApplicationCommand]
                                  │ RSSI: -88 dBm
                                  └─[NotificationCCReport]
                                      notification type:   Access Control
                                      notification status: 255
                                      notification state:  Window/door is open
                                      state parameters:    Window/door is open in regular position
2023-02-11T00:09:12.684Z CNTRLR   [Node 192] [~] [Notification] Access Control[Door state]: 23 => 5 [Endpoint 0]
                                  632
```

The "open in regular position" does flow back to an entity in HA, but that still isn't enough to make the "window/door is open" entity state change.  I've tried disabling the tilt sensor, but no change in behavior.  While the 1.x devices do all show in ZWaveJS-UI as having the tilt sensor enabled, I am not completely confident there even is one in the 1.x hardware because I can't get the 1.x sensors to tell me if they're opened normally or tilted.

Does this sound like something I should pursue here, or over on the HA side of the house?  Or given the difference in 1.x and 2.x hardware capabilities, maybe both?

Thanks.

**Comment by AlCalzone** [comment]
Have you tried re-interviewing the 2.x device? I very recently added support for parsing these kinds of state parameters, and the information that the value exists may not be stored yet.
If that doesn't help, I think this needs a fix on the HA side of things.

**Reply by jdiegmueller** [reply]
Yes, numerous times.  I do think the behavior changed recently but I confess I tinker so often I'm not confident which side changed behavior.

Do you know if the 1.x even has a tilt sensor?

**Reply by AlCalzone** [reply]
don't know tbh. There's no way to query for which subset of these "new" kinds of notification values a device has support for.

**Reply by neilbrencode** [reply]
I filled an HA issue about this when my Zwa012 stopped reporting clear/detected and started displaying raw state numbers for a related entity in the HA UI:
https://github.com/home-assistant/core/issues/88060

The HA contributor who took a look said HA expects ZWaveJS to provide labels for all the state values (like open in tilt position, open in regular position) and suggested I file a ZWaveJS issue.

**Reply by kpine** [reply]
There are two problems from the HA side.

First, there is a "string" sensor that just reflects the "raw" state of the notification, by mapping the raw value to its state label. There is no state key for the new values, so it can't translate from raw value to label.

Second, the binary sensors created for convenience are looking at the mapped "on" states. For a door sensor, "on" was previously only value 22. Now that the driver is reporting these new values (and only the new ones), they aren't recognized as "on" anymore. In that case, HA would need to add these new values in its notification mapping, but in addition these new states need to be available in value metadata.

**Reply by AlCalzone** [reply]
> but in addition these new states need to be available in value metadata

And they are after a re-interview.

**Reply by kpine** [reply]
Not according to https://github.com/home-assistant/core/issues/88060#issuecomment-1429286510. 🤷🏻‍♂️

I guess someone should upload an interview debug log.

**Reply by jdiegmueller** [reply]
> > but in addition these new states need to be available in value metadata
> 
> And they are after a re-interview.

Respectfully, there must be more to this then, as I have both re-interviewed and even exclude/re-included quite a bit since first digging into this over the weekend.

**Reply by jdiegmueller** [reply]
> Not according to https://github.com/home-assistant/core/issues/88060#issuecomment-1429286510. 🤷🏻‍♂️
> 
> I guess someone should upload an interview debug log.

I'll have this attached in a few minutes.  I'll do a reinterview of both a 1.x device followed by a 2.x device since the behavior differs.

Thanks guys.

**Reply by kpine** [reply]
@jdiegmueller Could you also attach (following re-interview) both the HA device diagnostic (device page -> ... -> Download diagnostic) and the ZUI node debug info?

**Reply by jdiegmueller** [reply]
Ok, hopefully I pulled it all together correctly.  Node 140 is "Outside Garage Wall Cabinet Door Sensor Left", Node 192 is "Outside Garage Wall Cabinet Door Sensor Right".  

zwave_js reinterviews:
[zwave_js-reinterviews.txt](https://github.com/zwave-js/node-zwave-js/files/10735549/zwave_js-reinterviews.txt)

zui:
[zui-debug_node140.txt](https://github.com/zwave-js/node-zwave-js/files/10735552/zui-debug_node140.txt)

[zui-debug_node192.txt](https://github.com/zwave-js/node-zwave-js/files/10735553/zui-debug_node192.txt)

homeassistant [UPDATED SINCE ORIGINAL POST]:
[zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside Garage Wall Cabinet Door Sensor Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt](https://github.com/zwave-js/node-zwave-js/files/10735649/zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside.Garage.Wall.Cabinet.Door.Sensor.Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt)

[zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside Garage Wall Cabinet Door Sensor Left-84fbf7a3e48cdbac562b764e86766b08.json.txt](https://github.com/zwave-js/node-zwave-js/files/10735652/zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside.Garage.Wall.Cabinet.Door.Sensor.Left-84fbf7a3e48cdbac562b764e86766b08.json.txt)

**Reply by kpine** [reply]
> homeassistant [UPDATED SINCE ORIGINAL POST]: [zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside Garage Wall Cabinet Door Sensor Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt](https://github.com/zwave-js/node-zwave-js/files/10735649/zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside.Garage.Wall.Cabinet.Door.Sensor.Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt)

~So what changed? Was the device not actually re-interviewed? The original file before the update did not have the new state keys. Now we see them.~ EDIT: I maybe have been looking at the wrong HA device file to being with, so ignore that.

According to the device diagnostic, after the re-interview successfully HA has picked up the new state keys, so that part is working as intended:

```json
        {
          "endpoint": 0,
          "commandClass": 113,
          "commandClassName": "Notification",
          "property": "Access Control",
          "propertyKey": "Door state",
          "propertyName": "Access Control",
          "propertyKeyName": "Door state",
          "ccVersion": 8,
          "metadata": {
            "type": "number",
            "readable": true,
            "writeable": false,
            "label": "Door state",
            "ccSpecific": {
              "notificationType": 6
            },
            "min": 0,
            "max": 255,
            "states": {
              "22": "Window/door is open",
              "23": "Window/door is closed",
              "5632": "Window/door is open in regular position",
              "5633": "Window/door is open in tilt position"
            }
          },
          "nodeId": 192,
          "value": 5632
        },
```

So 1) you will need to successfully re-interview affected devices and 2) HA needs to update to handle the new states. Based on the current design, you will get 3 different binary sensors. The old one, keying off of state 22 will probably never be functional (unless disabling tilt functionality disable the event parameters). The other two will only work when the tilt sensor is in the matching state.

**Reply by jdiegmueller** [reply]
> > homeassistant [UPDATED SINCE ORIGINAL POST]: [zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside Garage Wall Cabinet Door Sensor Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt](https://github.com/zwave-js/node-zwave-js/files/10735649/zwave_js-a9d099394b4d3da769910d186ccbb94b-Outside.Garage.Wall.Cabinet.Door.Sensor.Right-16ec0c4687a6a5a3755353fad384a3ce.json.txt)
> 
> So what changed? Was the device not actually re-interviewed? The original file before the update did not have the new state keys. Now we see them.
> 
> ```json
>         {
>           "endpoint": 0,
>           "commandClass": 113,
>           "commandClassName": "Notification",
>           "property": "Access Control",
>           "propertyKey": "Door state",
>           "propertyName": "Access Control",
>           "propertyKeyName": "Door state",
>           "ccVersion": 8,
>           "metadata": {
>             "type": "number",
>             "readable": true,
>             "writeable": false,
>             "label": "Door state",
>             "ccSpecific": {
>               "notificationType": 6
>             },
>             "min": 0,
>             "max": 255,
>             "states": {
>               "22": "Window/door is open",
>               "23": "Window/door is closed",
>               "5632": "Window/door is open in regular position",
>               "5633": "Window/door is open in tilt position"
>             }
>           },
>           "nodeId": 192,
>           "value": 5632
>         },
> ```

Yes I'm sorry, Ignore the first set from HA.  I have two sets of sensors and very similarly named, I downloaded/attached the wrong ones from HA on the first pass.  Node 140 and node 192 are what I should have originally provided.

I do have an entity in HA named "Window/door is open in regular position" and "Window/door is open in tilt position".  These entities do go "on" and "off".  But the "window/door is open" entity does not change on the Node 192 device.

**Reply by kpine** [reply]
> I do have an entity in HA named "Window/door is open in regular position" and "Window/door is open in tilt position". These entities do go "on" and "off".

Wow, I completely missed that. I was not expecting to see those, but now I see why that is the case (I think it's an accident). So it looks like everything is working as intended then.

> But the "window/door is open" entity does not change on the Node 192 device.

This sensor will never work, unless disabling tilt affects the reports. The device does not report value "22" any longer. You need to switch to the other entities, or try disabling tilt and see if that does anything.

**Reply by jdiegmueller** [reply]
> So 1) you will need to successfully re-interview affected devices and 2) HA needs to update to handle the new states. Based on the current design, you will get 3 different binary sensors. The old one, keying off of state 22 will probably never be functional (unless disabling tilt functionality disable the event parameters). The other two will only work when the tilt sensor is in the matching state.

I had the same thought re: disabling the tile sensor, but unfortunately it already is in what you see for Node 192.  I still get 5632 when the device is opened on the Node 192/device with 2.x firmware.  Node 140 doesn't do this.  I have a hunch it and the other 1.x firmware devices don't actually have tilt sensors.

From my user perspective, I can adapt to new entities if that's the way it goes.  But would it be possible to add device_class's to the entities?

**Reply by kpine** [reply]
The entity settings in HA let you choose what kind of device it is. HA can be updated to do this the same way as the standard sensor.

Your other option is to use the disabled-by-default binary sensor and update the device class the same way in the UI. Similarly, a "Door/Window" binary sensor could be set a default device class, it's just not implemented as these are "legacy" sensors.

HA used to create what I'll call "opposite" sensors, i.e. the closed state of 23 for a door sensor would be set as the sensor "on" state, and "not 23" would flip the entity state to off. These were removed at some point as they are usually confusing and redundant. Would be handy in this case though...

I imagine this is going to be confusing for other users going forward.

**Reply by neilbrencode** [reply]
I did try disabling tilt on this device and it returns 23 (closed) and 5632 (open in regular position)

**Reply by kpine** [reply]
One possibility of keeping the original sensor (open/close w/o caring about tilt) would be if HA were smart enough to mask the state value and see if it matches the "base" value. I don't know if HA devs would be up for something like that, esp. if Z-Wave JS doesn't provide the means of determining when this is possible.

**Reply by AlCalzone** [reply]
Unfortunately, this regular position/tilt information is encoded in a different way than most notification states (as an "enum" byte in the event parameters), and the Notification CC provides no way of checking whether a device supports reporting those, or only the "standard" states. I guess a heuristic could be to decide on the first incoming report, but even that's no 100% guarantee.

What makes it worse, the window status is the only instance where the enum values (partially) extend the defined values, instead of replacing them (which would make it easier).

**Reply by jdiegmueller** [reply]
> > I do have an entity in HA named "Window/door is open in regular position" and "Window/door is open in tilt position". These entities do go "on" and "off".
> 
> Wow, I completely missed that. I was not expecting to see those, but now I see why that is the case (I think it's an accident). So it looks like everything is working as intended then.
>
> > But the "window/door is open" entity does not change on the Node 192 device.
> 
> This sensor will never work, unless disabling tilt affects the reports. The device does not report value "22" any longer. You need to switch to the other entities, or try disabling tilt and see if that does anything.

re: disabling tilt sensor, that was my thought, too, but the tilt sensor is already disabled in the debugs I provided -- it didn't change the reporting behavior for this particular device.

Right or wrong, here's what I observe: ZWA012's running firmware 1.x (ie, will only ever report 22) result in two impossible entities (the entities representing 5622/5623) being created.  ZWA012's running firmware 2.x (ie will only ever report 5622 & 5623) result in one impossible entity (the entity representing 22).  

Speaking solely from the user POV here, my suggestion would be that entities which represent states that are impossible not get created, and that the new entities we are transitioning too get the correct device_class tagged on them.

Probably related: I also now see a motion sensor entity showing up in HA on Node 192?: Outside Garage Wall Cabinet Door Sensor Right Motion detection (location provided).  I do not see this in HA on Node 140.

**Reply by kpine** [reply]
I see this as analogous to the Door Look CC status variables, or the Battery CC variables, you just can't determine if the functionality is supported or not (depending on CC versions), so it's better to support it than not support it. The only real way to solve these are device and firmware version specific, which (as you know) both HA and Z-Wave JS try to avoid, for at least one reason being because it's a maintenance nightmare... The heuristic approach might be an issue for HA, which relies on state being already being known when it connects to the server, or gets the node "ready" events.

In this case though, I think HA could solve it by simply toggling the default binary sensor (for state 22) to "on" when `value >> 8 == state (22)`. However, there's no way to how and when to shift w/o knowing the implementation details.

**Reply by AlCalzone** [reply]
> entities which represent states that are impossible not get created

Which requires knowing that they are impossible. https://github.com/zwave-js/node-zwave-js/discussions/5453#discussioncomment-4974517

**Reply by AlCalzone** [reply]
> HA could solve it by simply toggling the default binary sensor (for state 22) to "on" when `value >> 8 == state (22)`. However, there's no way to how and when to shift w/o knowing the implementation details.

This isn't documented, but all enum values which share the same base event have the same high byte: 0x16, 0x1600, 0x1601. All 1-byte values are the actual notification events, the 2-byte ones are a combined event+enum. So a "window" open state can be determined by looking at the high byte, even if the value has a different byte size.

**Reply by jdiegmueller** [reply]
> Probably related: I also now see a motion sensor entity showing up in HA on Node 192?: Outside Garage Wall Cabinet Door Sensor Right Motion detection (location provided). I do not see this in HA on Node 140.

I have a better understanding that omitting the "impossible states" is a bigger problem than I realize with regards to open/closed/open regular/open tilt .. but what about this motion sensor bit I mentioned started showing up on the ZWA012 v2.x firmware devices?:
![image](https://user-images.githubusercontent.com/18660571/219040411-ee000c51-b098-4d59-8dbb-67d22efcc6ca.png)

![image](https://user-images.githubusercontent.com/18660571/219040706-5757eea6-21d5-4497-b7d2-a9cc5faf1df2.png)

**Reply by AlCalzone** [reply]
> but what about this motion sensor bit I mentioned started showing up on the ZWA012 v2.x firmware devices?

That's because the device reports that it supports event 7 for notification type 7, not event 3 like the manual claims:
![grafik](https://user-images.githubusercontent.com/17641229/219044160-6e926717-3d47-4cb8-b5ff-7e8b01636d40.png)

```
010e00a80001c005710207018000a7cf
                    ^^  ^^
                    ||  bit mask 0b1000_0000, bit 7 = Motion detection (location provided)
                    notification type 0x07 = Home Security
```

**Reply by jdiegmueller** [reply]
> > but what about this motion sensor bit I mentioned started showing up on the ZWA012 v2.x firmware devices?
> 
> That's because the device reports that it supports event 7 for notification type 7, not event 3 like the manual claims: ![grafik](https://user-images.githubusercontent.com/17641229/219044160-6e926717-3d47-4cb8-b5ff-7e8b01636d40.png)
> 
> ```
> 010e00a80001c005710207018000a7cf
>                     ^^  ^^
>                     ||  bit mask 0b1000_0000, bit 7 = Motion detection (location provided)
>                     notification type 0x07 = Home Security
> ```

Ok, cool.  I'll see if I can get Aeotec's attention on this, perhaps we can get it fixed in a future v2.x firmware.

Thanks again, everyone who weighed in here.

**Comment by jdiegmueller** [comment]
Thank you both for weighing in on this.  Appreciate it.

**Reply by matthijsvdr** [reply]
And I just ran into the same problem 😑 any chance you have an update ? 😄

**Comment by Jusdem** [comment]
Is this an issue with Aeotec firmware or an update to HA integration? Im also having this issue.
